CRM/Eck/BAO/EckEntityType.php

```diff
@@ -19,25 +19,22 @@
 
 class CRM_Eck_BAO_EckEntityType extends CRM_Eck_DAO_EckEntityType {
 
-  protected static $_entityTypes;
-
-  public static function getEntityTypes() {
-    if (!isset(static::$_entityTypes)) {
-      $entity_types = civicrm_api3('EckEntityType', 'get', [], ['limit' => 0])['values'];
-      static::$_entityTypes = array_combine(
-        array_map(
-          function ($name) {
-            return 'Eck' . $name;
-          },
-          array_column($entity_types, 'name')
-        ),
-        $entity_types
-      );
+  /**
+   * @return array[]
+   */
+  public static function getEntityTypes(): array {
+    if (!isset(Civi::$statics['EckEntityTypes'])) {
+      Civi::$statics['EckEntityTypes'] = (array) \Civi\Api4\EckEntityType::get(FALSE)
+        ->addSelect('*', 'CONCAT("Eck", name) AS entity_name')
```

Comment:
Instead, I added a column which concatenates `'Eck' . $name` and called it `entity_name`.

Reply:
Calling Api4 here (as opposed to Api3) causes an infinite loop when rebuilding cache because of the `civi.api4.entityTypes` event being repeatedly invoked following the `Entity::get()` Api4 call in `Civi\Api4\Utils\CoreUtil::getInfoItem()`:

```
EckEntityType.php:29, CRM_Eck_BAO_EckEntityType::getEntityTypes()
Entity.php:69, Civi\Eck\API\Entity->onApi4EntityTypes()
EventDispatcher.php:264, Symfony\Component\EventDispatcher\EventDispatcher->doDispatch()
EventDispatcher.php:239, Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
EventDispatcher.php:73, Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
CiviEventDispatcher.php:198, Civi\Core\CiviEventDispatcher->dispatch()
Get.php:51, Civi\Api4\Action\Entity\Get->getRecords()
BasicGetAction.php:52, Civi\Api4\Generic\BasicGetAction->_run()
ActionObjectProvider.php:68, Civi\Api4\Provider\ActionObjectProvider->invoke()
Kernel.php:149, Civi\API\Kernel->runRequest()
AbstractAction.php:234, Civi\Api4\Generic\AbstractAction->execute()
CoreUtil.php:67, Civi\Api4\Utils\CoreUtil::getInfoItem()
Api4SelectQuery.php:106, Civi\Api4\Query\Api4SelectQuery->__construct()
DAOGetAction.php:112, Civi\Api4\Generic\DAOGetAction->getObjects()
DAOGetAction.php:101, Civi\Api4\Generic\DAOGetAction->_run()
ActionObjectProvider.php:68, Civi\Api4\Provider\ActionObjectProvider->invoke()
Kernel.php:149, Civi\API\Kernel->runRequest()
AbstractAction.php:234, Civi\Api4\Generic\AbstractAction->execute()
EckEntityType.php:29, CRM_Eck_BAO_EckEntityType::getEntityTypes()
Entity.php:69, Civi\Eck\API\Entity->onApi4EntityTypes()
EventDispatcher.php:264, Symfony\Component\EventDispatcher\EventDispatcher->doDispatch()
EventDispatcher.php:239, Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
EventDispatcher.php:73, Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
CiviEventDispatcher.php:198, Civi\Core\CiviEventDispatcher->dispatch()
Get.php:51, Civi\Api4\Action\Entity\Get->getRecords()
BasicGetAction.php:52, Civi\Api4\Generic\BasicGetAction->_run()
ActionObjectProvider.php:68, Civi\Api4\Provider\ActionObjectProvider->invoke()
Kernel.php:149, Civi\API\Kernel->runRequest()
AbstractAction.php:234, Civi\Api4\Generic\AbstractAction->execute()
CoreUtil.php:67, Civi\Api4\Utils\CoreUtil::getInfoItem()
Api4SelectQuery.php:106, Civi\Api4\Query\Api4SelectQuery->__construct()
DAOGetAction.php:112, Civi\Api4\Generic\DAOGetAction->getObjects()
DAOGetAction.php:101, Civi\Api4\Generic\DAOGetAction->_run()
ActionObjectProvider.php:68, Civi\Api4\Provider\ActionObjectProvider->invoke()
Kernel.php:149, Civi\API\Kernel->runRequest()
AbstractAction.php:234, Civi\Api4\Generic\AbstractAction->execute()
ManagedEntities.php:663, CRM_Core_ManagedEntities->loadManagedEntityActions()
ManagedEntities.php:131, CRM_Core_ManagedEntities->reconcile()
Invoke.php:409, CRM_Core_Invoke::rebuildMenuAndCaches()
Invoke.php:100, CRM_Core_Invoke::hackMenuRebuild()
Invoke.php:65, CRM_Core_Invoke::_invoke()
Invoke.php:36, CRM_Core_Invoke::invoke()
Civicrm.php:88, Drupal\civicrm\Civicrm->invoke()
CivicrmController.php:80, Drupal\civicrm\Controller\CivicrmController->main()
EarlyRenderingControllerWrapperSubscriber.php:123, call_user_func_array:{/var/www/civicrm_env/drupal9/drupal/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:123}()
EarlyRenderingControllerWrapperSubscriber.php:123, Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure:/var/www/civicrm_env/drupal9/drupal/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:121-124}()
Renderer.php:564, Drupal\Core\Render\Renderer->executeInRenderContext()
EarlyRenderingControllerWrapperSubscriber.php:124, Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext()
EarlyRenderingControllerWrapperSubscriber.php:97, Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure:/var/www/civicrm_env/drupal9/drupal/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:96-98}()
HttpKernel.php:158, Symfony\Component\HttpKernel\HttpKernel->handleRaw()
HttpKernel.php:80, Symfony\Component\HttpKernel\HttpKernel->handle()
Session.php:58, Drupal\Core\StackMiddleware\Session->handle()
KernelPreHandle.php:48, Drupal\Core\StackMiddleware\KernelPreHandle->handle()
PageCache.php:106, Drupal\page_cache\StackMiddleware\PageCache->pass()
PageCache.php:85, Drupal\page_cache\StackMiddleware\PageCache->handle()
ReverseProxyMiddleware.php:48, Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle()
NegotiationMiddleware.php:51, Drupal\Core\StackMiddleware\NegotiationMiddleware->handle()
StackedHttpKernel.php:23, Stack\StackedHttpKernel->handle()
DrupalKernel.php:708, Drupal\Core\DrupalKernel->handle()
index.php:19, {main}()
```

I remember using Api3 here because of that as a "quick fix", which it obviously isn't in the long run. An entity type being dependent on another entity type seems to be a new situation that needs to be taken care of in Core somewhere, I guess. What do you think, @colemanw?

Reply:
In that case, I would say it should use a direct sql query.

Reply:
You're right, I didn't see the most obvious solution.